RE: Fix title and footer text within homepage shelves (isHomepageCard props)

xlisachan opened the pull request:
Fixes #10373 
Fixes #10509

@bobsilverberg, as discussed, this pull request fixes the overlap between the header and footer_text within the homepage shelves by passing through a props (`isHomepageCard`) that `Card` can then utilize. If we proceed with this update, I'll work on updating the tests.

Please see screenshots below:
Homepage (width: 1024px)-
<img width="765" alt="Screen Shot 2021-06-09 at 5 02 31 PM" src="https://user-images.githubusercontent.com/28129806/121429326-c7ae9a80-c944-11eb-9887-989efd1f622b.png">

Homepage (width: 768px)- 
<img width="764" alt="Screen Shot 2021-06-09 at 5 02 44 PM" src="https://user-images.githubusercontent.com/28129806/121429340-cc734e80-c944-11eb-8f9f-12841c170d70.png">

Homepage (width: 425px)- 
<img width="418" alt="Screen Shot 2021-06-09 at 5 03 00 PM" src="https://user-images.githubusercontent.com/28129806/121429348-d1380280-c944-11eb-95a8-47f447ec1fe2.png">

Rating (included as it utilizes `Card` component)-
<img width="270" alt="Screen Shot 2021-06-09 at 5 03 49 PM" src="https://user-images.githubusercontent.com/28129806/121429364-d5642000-c944-11eb-9052-d7f36b7efbb5.png">

Review (included as it utilizes `Card` component)-
<img width="762" alt="Screen Shot 2021-06-09 at 5 03 56 PM" src="https://user-images.githubusercontent.com/28129806/121429379-da28d400-c944-11eb-924b-086ed15190bb.png">

Please review when you have a moment. Thank you!

xlisachan commented:
@bobsilverberg - per your feedback, I have updated the tests for this approach. The latest push also includes an updated Card stylesheet as the [at]extend rule was failing CI tests. Please see screenshots below for reference.

Homepage (width: 1024)
<img width="765" alt="Screen Shot 2021-06-16 at 3 14 17 PM" src="https://user-images.githubusercontent.com/28129806/122279267-da315280-ceb5-11eb-801b-01d2b4c90ec6.png">

Homepage (width: 768)
<img width="761" alt="Screen Shot 2021-06-16 at 3 14 27 PM" src="https://user-images.githubusercontent.com/28129806/122279278-de5d7000-ceb5-11eb-9eee-5adb235f1a0f.png">

Homepage (width: 425)
<img width="420" alt="Screen Shot 2021-06-16 at 3 14 36 PM" src="https://user-images.githubusercontent.com/28129806/122279376-f8974e00-ceb5-11eb-89ec-d8eb6130b268.png">

Rating (as it utilizes Card)
<img width="294" alt="Screen Shot 2021-06-16 at 3 14 53 PM" src="https://user-images.githubusercontent.com/28129806/122279396-fe8d2f00-ceb5-11eb-9d43-ab935e866fdf.png">

Review (as it utilizes Card)
<img width="750" alt="Screen Shot 2021-06-16 at 3 15 02 PM" src="https://user-images.githubusercontent.com/28129806/122279415-02b94c80-ceb6-11eb-8589-051edd180754.png">

Please review when you have a moment. Thank you!

bobsilverberg commented on src/amo/components/Card/index.js:
```diff
@@ -63,12 +66,34 @@ export default class Card extends React.Component {
+            <div className="Card-header-text">{header}</div>
+
+            {isHomepageCard && footer ? (
+              <footer className="HomepageCard-footer-in-header">
```
Because this is in the `Card` component, the `className` should start with `Card` and not `HomepageCard`.

xlisachan commented on src/amo/components/AddonsCard/index.js:
```diff
@@ -44,6 +45,7 @@ type Props = {|
   footerLink?: Object | string | null,
   footerText?: string,
   header?: React.Node,
+  isHomepageCard?: boolean,
```
I have updated this to be kept in `DefaultProps` and removed it from `Props`. I'm trying to remember if I was using `showPromotedBadge?`as an example - it was placed in both `DefaultProps` and `Props` as well (line 27 and line 40). Should I remove `showPromotedBadge?` in `Props`?

bobsilverberg replied:
I think `isHomepageCard` could be just in `Props` and not in `DefaultProps`, because it will default to `false` if omitted. I think the reason `showPromotedBadge` is in both places is because it is defaulted to `true` when `defaultProps` are set.

bobsilverberg replied:
Actually, I take all of that back! They can both stay in `DefaultProps` and be removed from `Props`.

xlisachan commented:
Thanks for the feedback, @bobsilverberg !

The latest push includes all of the updates. Please let me know how to proceed with `AddonsCard` component (see comment regarding `showPromotedBadge?` property above).

Please review when you have a moment. Thank you!

bobsilverberg commented on src/amo/components/AddonsCard/index.js:
```diff
@@ -19,6 +19,7 @@ import './styles.scss';
 
 type DefaultProps = {|
   editing?: boolean,
+  isHomepageShelf?: boolean,
```
As discussed in my reply to your question, I think this should be moved into `Props`.

bobsilverberg commented on src/amo/components/AddonsCard/index.js:
```diff
@@ -58,6 +59,7 @@ type Props = {|
 export default class AddonsCard extends React.Component<Props> {
   static defaultProps: DefaultProps = {
     editing: false,
+    isHomepageShelf: false,
```
This can be removed.

bobsilverberg commented on src/amo/components/Card/index.js:
```diff
@@ -13,10 +13,12 @@ export default class Card extends React.Component {
   };
 
   static defaultProps = {
+    isHomepageShelf: false,
```
This isn't needed.

bobsilverberg commented on tests/unit/amo/components/TestLandingAddonsCard.js:
```diff
@@ -63,6 +63,18 @@ describe(__filename, () => {
     );
   });
 
+  it.each([true, false])(
```
You marked this as done, but I still see this testing both `true` and `false`.

xlisachan replied:
Ah, I had updated the wrong test - sorry! Updated!

bobsilverberg commented on tests/unit/amo/components/TestAddonsCard.js:
```diff
@@ -166,6 +166,18 @@ describe(__filename, () => {
     expect(root.find(CardList)).toHaveProp('header', header);
   });
 
+  it.each([true, false])(
```
I missed this last time, but as with the test for `LandingAddonsCard`, we don't really need to test both `true` and `false` here. I'd just test with `true`.

xlisachan replied:
Ah, I had updated the wrong test - sorry! Updated!

bobsilverberg reviewed (changes requested):
Looks good, thanks for the updates @xlisachan. I answered your question and added a couple of comments.

xlisachan commented:
Thanks @bobsilverberg !

I removed the `showPromotedBadge?` property from `Props` as advised, and updated the correct tests - sorry about that!

Please review when you have a moment. Thanks!

bobsilverberg reviewed (changes requested):
It all looks good, thanks for all of your efforts @xlisachan, great work!

I tried running it locally and the only issue I have is that it seems like the minimum gap between the shelf title and the footer link should be a bit wider. Here's an example of the home page on dev at 758px wide:

![Screen Shot 2021-07-19 at 1 55 00 PM](https://user-images.githubusercontent.com/142755/126204934-3ec841b4-5e7f-4c9f-b533-0f8e3606789e.png)

Can you see if you can specify a minimum gap so that there is still a bit of space between them in that scenario?

xlisachan commented:
I added a `margin-left` property to `.Card-shelf-footer-in-header` as adding a margin to the `.Card-header-text` would affect non-Homepage Cards. Please let me know if you're aligned and/or whether `16px` is enough space.

Screen width: 774px
<img width="751" alt="Screen Shot 2021-07-20 at 3 59 37 PM" src="https://user-images.githubusercontent.com/28129806/126387797-140e1b9e-fdca-45fe-aebb-3c0550c8655a.png">

(w/ grid for reference)
<img width="762" alt="Screen Shot 2021-07-20 at 3 59 32 PM" src="https://user-images.githubusercontent.com/28129806/126387818-1455f889-4243-4958-bdf5-40b2196da9dc.png">

Please review when you have a moment. Thanks!

bobsilverberg reviewed (approved):
It looks great, thanks @xlisachan.

This is an issue that has been around for a long time, and a fix has been attempted by contributors in the past, but nobody was successful. Thank you for sticking with it and coming up with a solution that works so well. Well done @xlisachan!
